robot/web/web.go

```diff
@@ -558,15 +571,77 @@ type requestStats struct {
 	dataSent  atomic.Int64
 }
 
-// RequestCounter maps string keys to atomic ints that get bumped on every incoming gRPC request for
-// components.
+type requestCounter struct {
+	mu   sync.Mutex
+	curr int
+	max  int
+}
+
+// requestLimiter is used to track and limit the number of concurrent requests
+// for resources. By default the per-resource limit is 100 concurrent requests
+// but this can be changed by setting the `VIAM_RESOURCE_REQUESTS_LIMIT`
+// environment variable.
+type requestLimiter struct {
+	counts ssync.Map[string, *requestCounter]
+	limit  int
+}
+
+func (l *requestLimiter) ensureLimit() {
+	if l.limit == 0 {
+		if limitVar, err := strconv.Atoi(os.Getenv("VIAM_RESOURCE_REQUESTS_LIMIT")); err == nil && limitVar > 0 {
+			l.limit = limitVar
+		} else {
+			l.limit = 100
+		}
+	}
+}
+
+func (l *requestLimiter) ensureKey(resource string) *requestCounter {
+	counter, ok := l.counts.Load(resource)
+	if !ok {
+		counter, _ = l.counts.LoadOrStore(resource, &requestCounter{})
+	}
+	return counter
+}
+
+// Incr attempts to increment the in-flight request counter for a given
+// resource. It returns true if it was successful and false if an additional
+// request would exceed the configured limit.
+func (l *requestLimiter) Incr(resource string) bool {
+	counter := l.ensureKey(resource)
+	counter.mu.Lock()
+	if counter.curr >= l.limit {
+		counter.mu.Unlock()
+		return false
+	}
+	counter.curr++
+	counter.mu.Unlock()
+	// Intentionally leave this outside the critical section so we can overwrite
+	// it in Stats() without the race detector failing our tests.
+	counter.max = max(counter.curr, counter.max)
```

Comment:
I intentionally left a race on the variable tracking the max concurrent connections because I assume it `Stats()` we don't want `Stats` to take a bunch of locks as it's trying to flush stats to disk every second. This means the stats may be slightly off within each window but it should be correct enough to be useful. If we want it to be precise and it's ok to take locks in `Stats()` I can update the code to always take the lock.

Reply:
I don't see how the race detector would allow reading `counter.curr` outside of the lock.

Similar I don't see how the race detector would allow writing to `counter.max` in this goroutine while the FTDC thread can simultaneously read from it.